llvm/lib/Target/RISCV/RISCVInstrInfo.td

```diff
@@ -493,7 +493,15 @@ def LeadingOnesMask : PatLeaf<(imm), [{
   if (!N->hasOneUse())
     return false;
   return !isInt<32>(N->getSExtValue()) && isMask_64(~N->getSExtValue());
-}], TrailingZeros>;
+}], TrailingZeros> {
+  let GISelPredicateCode = [{
+    if (!MRI.hasOneNonDBGUse(MI.getOperand(0).getReg()))
+      return false;
+    const auto &MO = MI.getOperand(1);
```

Comment:
@topperc i presume these are looking at different operand indexes because `MI` is a `G_CONSTANT` (or `G_FCONSTANT`)? or is something else going on here. What validates that this MI is actually what is being expected? I looked through some of the code, but definitely was out of my depth (and I don't think the other GISelPredicates seem to check this either.

Reply:
GISel uses MachineIR where defs and uses are "operands". For G_CONSTANT, operand 0 is the def and operand 1 is the source.

SelectionDAG operands are only inputs. Results are separate.

The `(imm)` in the PatLeaf should guarantee this is only called for G_CONSTANT.

Reply:
> GISel uses MachineIR where defs and uses are "operands". For G_CONSTANT, operand 0 is the def and operand 1 is the source.

Ok, so `MI` is the instruction that "defines" the vreg being used in the instruction being matched. And we have to look at its operands because immediates are not stored inline (except in G_CONSTANT). That makes sense.

> The (imm) in the PatLeaf should guarantee this is only called for G_CONSTANT

I took a look through the generated match table code, and I couldn't find what guaranteed this, which is part of where my confusion came from.

The handler for `GIM_CheckCxxInsnPredicate` doesn't check the opcode before calling `testMIPredicate_MI` (which contains this C++), and if you look at e.g. `MatchTable0` in `RISCVGenGlobalISel.inc`, there's not always a `GIM_CheckOpcode` with `G_CONSTANT` before `GIM_CheckCxxInsnPredicate`.

Reply:
> > The (imm) in the PatLeaf should guarantee this is only called for G_CONSTANT
> 
> I took a look through the generated match table code, and I couldn't find what guaranteed this, which is part of where my confusion came from.
> 
> The handler for `GIM_CheckCxxInsnPredicate` doesn't check the opcode before calling `testMIPredicate_MI` (which contains this C++), and if you look at e.g. `MatchTable0` in `RISCVGenGlobalISel.inc`, there's not always a `GIM_CheckOpcode` with `G_CONSTANT` before `GIM_CheckCxxInsnPredicate`.

GIM_CheckCxxInsnPredicate can be used for other opcodes than G_CONSTANT. The opcode check should be based on the root specified in the `PatLeaf` or `PatFrag`. In this case it was G_CONSTANT. For `GICXXPred_MI_Predicate_add_non_imm12` and , the opcode is G_ADD which is checked by the top level `GIM_SwitchOpcode` instead of a `GIM_CheckOpcode`.

Reply:
Thanks for the explanations, now I see what's going on.